File: ruby/run-length-encoding/run_length_encoding.rb
# frozen_string_literal: true

class RunLengthEncoding
  def self.encode(unencoded)
    return '' if unencoded.empty?

    encoded = []
    current = unencoded[0]
    count = 0
    unencoded.each_char do |character|
      if character == current
        count += 1
      else
        encoded << count.to_s
        encoded << current
        current = character
        count = 1
      end
    end
    encoded << count.to_s
    encoded << current
    encoded.keep_if { |token| token != '1' }
    encoded.join
  end

  def self.decode(encoded)
    return '' if encoded.empty?

    seq = encoded.split(/([0-9]+)/)
    seq.delete('')
    seq.map! { |elem| elem.match(/\A[0-9]+\z/) ? elem.to_i : elem.chars }
    seq.flatten!
    decoded = String.new
    seq.each.with_index do |token, index|
      if token.is_a?(Integer)
        decoded << (seq[index + 1] * token)
      elsif index.zero? || (index != 0 && !seq[index - 1].is_a?(Integer))
        decoded << token
      end
    end
    decoded
  end
end

File: java/run-length-encoding/RunLengthEncoding.java
import java.util.Arrays;
import java.util.Scanner;
import java.util.ArrayList;

public class RunLengthEncoding {
    public String encode(String unencoded) {
        if (unencoded.isEmpty()) {
            return unencoded;
        }

        Scanner scanner = new Scanner(unencoded);
        ArrayList<String> seq = new ArrayList<>();
        char current = unencoded.charAt(0);
        int count = 0;
        for (char character: unencoded.toCharArray()) {
            if (character == current) {
                count += 1;
            }
            else {
                seq.add(String.valueOf(count));
                seq.add(String.valueOf(current));
                current = character;
                count = 1;
            }
        }
        seq.add(String.valueOf(count));
        seq.add(String.valueOf(current));
        StringBuilder encoded = new StringBuilder();
        seq.forEach(s -> { if (!s.equals("1")) encoded.append(s); });

        return encoded.toString();
    }

    public String decode(String encoded) {

        // here be dragons
        encoded = encoded.replaceAll(
            "(\\p{Digit}|\\p{IsAlphabetic}| )( |\\p{IsAlphabetic})", "$1\0$2");
        encoded = encoded.replaceAll(
            "(\\p{IsAlphabetic}| )( |\\p{IsAlphabetic}|\\p{Digit})", "$1\0$2");
        encoded = encoded.replaceAll(" (\\p{Digit})", " \0$1");
        encoded = encoded.replaceAll(
            "(\\p{IsAlphabetic})(\\p{Digit}|\\p{IsAlphabetic}| )", "$1\0$2");

        String[] seq = encoded.split("\0");

        StringBuilder decoded = new StringBuilder();
        for (int i = 0; i < seq.length; i++) {
            String token = seq[i];
            try {
                int quantifier = Integer.parseInt(token);
                token = seq[++i];
                decoded.append(token.repeat(quantifier));
            }
            catch (NumberFormatException ex) {
                decoded.append(token);
            }
        }
        return decoded.toString();
    }
}
